cert/cert.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2012, 2013 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cert
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"time"
+
+	"github.com/juju/errors"
+)
+
+var KeyBits = 2048
+
+// ParseCert parses the given PEM-formatted X509 certificate.
+func ParseCert(certPEM string) (*x509.Certificate, error) {
+	certPEMData := []byte(certPEM)
+	for len(certPEMData) > 0 {
+		var certBlock *pem.Block
+		certBlock, certPEMData = pem.Decode(certPEMData)
+		if certBlock == nil {
+			break
+		}
+		if certBlock.Type == "CERTIFICATE" {
+			cert, err := x509.ParseCertificate(certBlock.Bytes)
+			return cert, err
+		}
+	}
+	return nil, errors.New("no certificates found")
+}
+
+// ParseCertAndKey parses the given PEM-formatted X509 certificate
+// and RSA private key.
+func ParseCertAndKey(certPEM, keyPEM string) (*x509.Certificate, *rsa.PrivateKey, error) {
+	tlsCert, err := tls.X509KeyPair([]byte(certPEM), []byte(keyPEM))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cert, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	key, ok := tlsCert.PrivateKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, nil, fmt.Errorf("private key with unexpected type %T", key)
+	}
+	return cert, key, nil
+}
+
+// Verify verifies that the given server certificate is valid with
+// respect to the given CA certificate at the given time.
+func Verify(srvCertPEM, caCertPEM string, when time.Time) error {
```

Comment:
This is not needed for winrm, winrm itself has built it function to verify the cert.